lib/kvdb/kvdb_ctxn.c

```diff
@@ -179,6 +198,16 @@ kvdb_ctxn_reaper(struct work_struct *work)
         kvdb_ctxn_abort(&ctxn->ctxn_inner_handle);
         ctxn->ctxn_expired = true;
     }
+    // The above code could be in a race with the client, so a txn being on the abort
+    // list doesn't guarantee it will be expired. Thus:
+    //   1. the "Aborting %u transactions" log message might be misleading.
+    //   2. ctxn_expired may end up being set on txn committed or aborted by the client.
+    // Fix:
+    // - Change 'kvdb_ctxn_abort()' to return true if and only if transitioned txn to aborted state.
+    // - Add a "reason" param to kvdb_ctxn_abort() with reason set to "expired".
+    //
+    // Should the reaper reset ctxn->ctxn_abort_link? I don't think it's
+    // necessary, but it seems more hygienic.
```

Comment:
I don't think it's a race, it's intentional.  But the "expired" stuff at line 180 doesn't look right.
Sigh, I'm going to have to review the entire thing...

Reply:
I mean, it used to be if the reaper lost the race then the ctxn would not be aborted because presumably the owner was now using it and dealing with it.  Now I just don't know...

Reply:
Intentional or not, I think this is true:
1. just because a txn is on the abort list `alist`, doesn't mean it will be aborted by this thread.  the client may get in there and change the txn state before this thread has a chance to do the abort.
2. as a result of 1, we might set expired=true on txns that weren't actually expired.

Reply:
So here is a silly question. Why do the transactions get moved to this abort list? Why not just abort them in that initial for each loop?

Reply:
I guess maybe to separate the writing from the reading? I see the read lock is taken while iterating the first for each loop

Reply:
Or to reduce the work while read lock is held.  Note also each txn abort requires a mutex on the ctxn lock.

Reply:
Because the abort() path is long and nasty and might sleep, and we don't want to do that while in the midst of an RCU read-side critical section.

Reply:
Probably that ctxn_expired = true should be moved into the abort function.  But I need to review..

Reply:
I think what Alex proposed about abort() taking a bool expired or some such is probably correct.

Reply:
Why can't abort just set ctnx_expired?

Reply:
How would you tell the difference from a user aborting a txn and the reaper thread doing it?

Reply:
oh, is see, you want that condition to persist for future notification, makes sense...

Reply:
Right. That boolean is the only thing which can alert the user the transaction expired

Reply:
I think abort should take an enum of some sort to indicate reason for aborting.  We can feed that back to the user if they try to use an invalid txn (eg: cursor read on txn that they aborted,  cursor seek on txn that expired).